File: solutions/javascript/captains-log/1/captains-log.js
// @ts-check

/**
 * Generates a random starship registry number.
 *
 * @returns {string} the generated registry number.
 */
export function randomShipRegistryNumber() {
  const randoms = crypto.getRandomValues(new Uint8Array(4))
  return `NCC-${randoms[0] % 9 + 1}${randoms[1] % 10}${randoms[2] % 10}${randoms[3] % 10}`
}

/**
 * Generates a random stardate.
 *
 * @returns {number} a stardate between 41000 (inclusive) and 42000 (exclusive).
 */
export function randomStardate() {
  const random = crypto.getRandomValues(new Uint32Array(2))
  const mantissa = ((random[0] / 2 ** 6) * 2 ** 26) + (random[1] / 2 ** 6)
  return 41000 + 1000 * mantissa / (2 ** 52)
}

/**
 * Generates a random planet class.
 *
 * @returns {string} a one-letter planet class.
 */
export function randomPlanetClass() {
  const PLANETS = ['D', 'H', 'J', 'K', 'L', 'M', 'N', 'R', 'T', 'Y']
  return PLANETS[crypto.getRandomValues(new Uint8Array(1))[0] % PLANETS.length]
}

File: solutions/java/captains-log/1/src/main/java/CaptainsLog.java
import java.util.random.RandomGenerator;

class CaptainsLog {
    private static final char[] PLANET_CLASSES = new char[]{'D', 'H', 'J', 'K', 'L', 'M', 'N', 'R', 'T', 'Y'};

    private RandomGenerator random;

    CaptainsLog(RandomGenerator random) {
        this.random = random;
    }

    char randomPlanetClass() {
        return PLANET_CLASSES[random.nextInt(PLANET_CLASSES.length)];
    }

    String randomShipRegistryNumber() {
        return "NCC-" + Integer.toString(random.nextInt(1000, 10000));
    }

    double randomStardate() {
        return random.nextDouble(41000.0, 42000.0);
    }
}